oximeter/db/src/oxql/query/mod.rs

```diff
@@ -1006,4 +1035,147 @@ mod tests {
+        let silo_id = Uuid::new_v4();
+        let project_id = Uuid::new_v4();
+        let new_query = query.add_filters(vec![
+            uuid_eq_filter("silo_id".to_string(), silo_id),
+            uuid_eq_filter("project_id".to_string(), project_id),
+        ]);
+
+        assert_eq!(query.parsed.table_ops().len(), 2);
+        assert_eq!(new_query.parsed.table_ops().len(), 4);
+
+        // inserted after the get
+        assert_eq!(
+            new_query.parsed.table_ops().nth(1).unwrap().to_string(),
+            format!("filter (silo_id == \"{}\")", silo_id)
+        );
+        assert_eq!(
+            new_query.parsed.table_ops().nth(2).unwrap().to_string(),
+            format!("filter (project_id == \"{}\")", project_id)
+        );
+    }
+
+    #[test]
+    fn test_add_filters_with_subqueries() {
+        let query =
+            Query::new("{ get a:b | filter timestamp > @2025-03-05; get c:d }")
+                .unwrap();
+        let silo_id = Uuid::new_v4();
+        let project_id = Uuid::new_v4();
+
+        // Define expected filters as AST nodes
+        let silo_filter = uuid_eq_filter("silo_id".to_string(), silo_id);
+        let project_filter =
+            uuid_eq_filter("project_id".to_string(), project_id);
+
+        let expected_silo_op =
+            TableOp::Basic(BasicTableOp::Filter(silo_filter.clone()));
+        let expected_project_op =
+            TableOp::Basic(BasicTableOp::Filter(project_filter.clone()));
+
+        let new_query = query.add_filters(vec![silo_filter, project_filter]);
+
+        // Check top-level structure (should remain a single grouped op)
+        assert_eq!(query.parsed.table_ops().len(), 1);
+        assert_eq!(new_query.parsed.table_ops().len(), 1);
+
+        let only_op = new_query.parsed.first_op();
+        let TableOp::Grouped(GroupedTableOp { ops }) = only_op else {
+            panic!("Expected the only operation to be TableOp::Grouped");
+        };
+
+        assert_eq!(ops.len(), 2, "Expected two subqueries in the group");
+
+        // first subquery has the original get and filter and now two extra filters in the middle
+        let subq1: Vec<_> = ops[0].table_ops().cloned().collect();
+        assert_eq!(subq1.len(), 4,);
+        assert_matches!(subq1[0], TableOp::Basic(BasicTableOp::Get(_)));
+        assert_eq!(subq1[1], expected_silo_op);
+        assert_eq!(subq1[2], expected_project_op);
+        assert_matches!(subq1[3], TableOp::Basic(BasicTableOp::Filter(_)));
+
+        // second subquery has the original get and now two extra filters
+        let subq2: Vec<_> = ops[1].table_ops().cloned().collect();
+        assert_eq!(subq2.len(), 3);
+        assert_matches!(subq2[0], TableOp::Basic(BasicTableOp::Get(_)));
+        assert_eq!(subq2[1], expected_silo_op);
+        assert_eq!(subq2[2], expected_project_op);
+    }
+
+    #[test]
+    fn test_add_filters_with_nested_subqueries() {
```

Comment:
Strictly speaking this test covers the case in the previous one, but it's also impossible to understand, so I'm ok with keeping both.

Reply:
More tests is better, keep them both.